json/json.go

```diff
@@ -67,25 +68,32 @@ func (s *jsonSearcher) jsonSearch(w http.ResponseWriter, req *http.Request) {
 		searchArgs.Opts = &zoekt.SearchOptions{}
 	}
 
-	query, err := query.Parse(searchArgs.Q)
+	q, err := query.Parse(searchArgs.Q)
 	if err != nil {
 		jsonError(w, http.StatusBadRequest, err.Error())
 		return
 	}
 
+	// TODO: We need to distinguish empty array from unset. An empty array
```

Comment:
I'm actually not sure how to check if an array is unset easily in Go because once I parse the JSON I lose the original value and just get empty array. I figured there might be some kind of hacky way to implement a custom JSON parser where it sets some value like `[333,333,333]` which we know is almost certainly not what a client sent in the request so we can assume once decoded that this was actually an empty value. Is there a better way to do this, though?

Reply:
> The reason at Sourcegraph we haven't needed to update the parser is we directly contstruct the query types and gob encode.

In our case GitLab will be constructing the list of `repoid` to filter by and we don't want this exposed to the user. In our case the ability to select projects (or groups of projects) is a dropdown selector and not part of the query syntax they use. Most importantly, however, the list of `repoid` being filtered is actually part of our security model because users only are allowed to search projects they are authorized to view.

Originally I started implementing this with prepending `repo:` to the filter because this API didn't exist but I quickly learnt that the user can escape our security model by just adding `or repo:/someprivaterepo` to the start of their query. There will be ways to protect against this with an allowlist of characters but ultimately the only fully correct way to protect from the GitLab side while still allowing all the query syntax flexibility for users would be to implement a parser in GitLab. Since GitLab is written in Rails this would just be impractical to implement a parser that matches the parser implemented in Zoekt.

So ultimately in the end the separate field as I implemented here was just the simplest option for us. I can certainly see the advantage of extending the query parser rather than adding new fields to the HTTP API but it does make it trickier for our workload where this field really needs to be treated differently from the rest of the query string.

Reply:
Ok I think I understand the limitations of this API and why you are doing it this way. The problem is you don't have a way to parse the users query, then manipulate it, and then output the query again. Indeed the zoekt/query package does not have a way to output an equivalent query string.

The reason Sourcegraph does not have that issue is we directly manipulate a `query.Q` and users don't have direct control. We use the same model you are doing for privacy. In sourcegraph we first find out the set of repos a user is allowed to view. However, we then construct the query `(repo_id_list AND user_query)`. This is what I am suggesting you do, but indeed you can't because your input to the RPC is a query string.

Note: For performance we actually do a little more. We rewrite queries to be look effectively like this `(private_repo_id_list AND user_query) or (repo_is_public AND user_query)`. IE we also store in zoekt if a repo is public, which allows us to bypass passing around huge lists of repos. This works with the assumption that the list of public repos is large while the list of private repos is relatively small in comparison.